### ISSUE TITLE
Engine/Compiler: Make order of scripts immaterial

### ISSUE THREAD

**Issue by fernewelten** [adventuregamestudio/ags#1387]
Currently, every script may only reference `import` variables and functions that have already been created in earlier scripts. So the order of scripts is important in this respect. 

Let's remove this restriction. The scripts should work no matter in which order they are in the Editor. 

AFAICS we would need to address two areas:

1. In the Engine, instances from scripts are built in the following way (pseudo code):
```
for (each script)
{
    {
        create an instance and allocate imports;
        resolve references to external imports; 
        resolve 'IMPORT' fixups;
    }
    fork off another instance for 'repeatedly_execute_always';
}
```

Let's reorder these steps: Create __all__ the instances first, only __then__ resolve references between instances:
```
for (each script)
    create an instance and allocate imports;

for (each script)
{
    resolve references to external imports; 
    resolve 'IMPORT' fixups;
}        

for (each script)
    fork off another instance for 'repeatedly_execute_always';
```

Now the sequence of scripts is immaterial in so far as functions are concerned. All scripts (room script, global script, user scripts) can call all the functions of all the scripts. The compiler doesn't need to be modified for this to work.

I've implemented this experimentally here: https://github.com/fernewelten/ags/tree/Engine_TwoPassLinking

2. The compiler must currently know a type definition before it can use it. 
 
So the let's change this and let __all__ enumeration and struct definitions be valid in the whole program instead of only from the point where they are declared. (This would implement a suggestion that @ivan-mogilko had some weeks ago.)

We'd need to reorder the two passes that the (new) compiler makes over the source code, as follows:
```
read through the source code and only process 'struct' and 'enum' definitions and function headers;
read through the source code and process the rest (variables with non-local scope, function bodies);
```

I'm going to implement that experimentally next.

-----

Open issues: 
* There are some event functions that each script may have individually, such as `game_start()`. I haven't found out yet whether these will be impacted by the Engine reordering
* This idea might make it possible to modularize GUIs better. Currently, GUI event functions must be in `GlobalScript`. When all the scripts are first loaded, then resolved, then GUI event functions might also work when in user script modules. I haven't explored that yet, there might be snags.

Do you have any insights to these? 
Have I overlooked some important aspect?

**Comment by ericoporto** [comment]
> There are some event functions that each script may have individually, such as game_start()

The order of these I think only matter at link-time in the engine.

I think event handlers have to be ordered, because this is how you can claim an event and prevent it to propagate. Repeatedly execute, game start and all these also happen in order - one may create some data at game_start and need it in a following game_start from other script, same for rep_exec.

> Currently, GUI event functions must be in GlobalScript

I am not sure the editor linking to user functions like for the GUI is related to this.

---

The auto-complete needs to be adjusted since it currently is aware of order - this may be other issue since there are currently other things out-of-sync with auto-complete when using the new compiler.

**Comment by fernewelten** [comment]
I haven't found any linking code in the Editor altogether. I might be wrong. But it looks as though the Editor simply bundles everything it has into the game file and shoves the lot to the Engine with very little modification. Then when the game is firing up, the Engine unpacks what it was given and resolves all external the symbols. 

In other words, all the linking seems to happen at begin of runtime, not as the concluding step of the compile time.

**Comment by ericoporto** [comment]
Yes, it's at runtime! I am not sure, but I think it links GUI to Global Script in a similar way the room stuff links to room script. I was trying to follow around GUI buttons and found this:

https://github.com/adventuregamestudio/ags/blob/5c6a1f2c613264b61b4fb0d7b8a6c6dc577c4677/Common/gui/guiobject.h#L107-L115

Maybe it needs an additional parameter to store in which script the function is?

**Comment by ivan-mogilko** [comment]
To summarize, there are two reasons the order of scripts matter in AGS now:

1. It defines which headers are included into scripts. The script gets headers only from above included automatically, but never from below.
2. It defines the order of event callbacks. These are called in the order of the scripts in the list. This is important, because the game logic may depend on that.


When it comes to importing declarations, dropping reliance on "physical" order of the items in the list may be a good thing, and also make it simplier to structure and build game projects in the future.
My concern is that there will be no immediate alternative for defining script dependency. That may lead to a less organized code, as it will be much easier to create unexpected ties and circular dependencies between scripts.
Therefore my proposal would be to consider making script dependency explicit. This may be done C-style (`#include header`) or C#/Java style (`using script`, `import script`).


Ordering callbacks is more essential, as a wrong order may break things in games. Unlike the previous problem, this cannot be tolerated temporarily, so the script order will have to define an order of callbacks until the alternative is implemented.
One of the potential solutions seem to be delegates (#1409), as afaik they define an order of calls (= the order the functions were subscribed). But that would require not only implementing delegates, but also giving a straightforward enough way to bind callbacks to game events. Right now it's very easy: you just write a function of certain name inside a script. I think the new solution must be as simple.

**Comment by ivan-mogilko** [comment]
To summarize, there are two reasons the order of scripts matters in AGS now:

1. It defines which headers are included into scripts. The script gets headers only from above included automatically, but never from below.
2. It defines the order of event callbacks. These are called in the order of the scripts in the list. This is important, because the game logic may depend on that.


When it comes to importing declarations, dropping reliance on "physical" order of the items in the list may be a good thing, and also make it simplier to structure and build game projects in the future.
My concern is that there will be no immediate alternative for defining script dependency. That may lead to a less organized code, as it will be much easier to create unexpected ties and circular dependencies between scripts.
Therefore my proposal would be to consider making script dependency explicit. This may be done C-style (`#include header`) or C#/Java style (`using script`, `import script`).


Ordering callbacks is more essential, as a wrong order may break things in games. Unlike the previous problem, this cannot be tolerated temporarily, so the script order will have to define an order of callbacks until the alternative is implemented.
One of the potential solutions seem to be delegates (#1409), as afaik they define an order of calls (= the order the functions were subscribed). But that would require not only implementing delegates, but also giving a straightforward enough way to bind callbacks to game events. Right now it's very easy: you just write a function of certain name inside a script. I think the new solution must be as simple.

**Comment by ericoporto** [comment]
An idea to the order of callbacks, not sure how feasible but, what about a "priority" decorator? Either `@priority(5)` or `[priority 5]` and so on, and use the number as a kind of "zorder-like" way of calling the stuff, so if you want to be explicit you do something like this to specify what needs to happen first - and maybe support negative priorities too,  in case we want to support script modules that may not have defined a priority. The reason I say decorator and not attribute is because the priority "level" would be a parameter and I don't know if those are possible in attributes - [solidity language](https://docs.soliditylang.org/en/v0.8.7/contracts.html) allows parameters in attributes by making them happen after the parameter list of the function.

**Comment by ivan-mogilko** [comment]
> An idea to the order of callbacks, not sure how feasible but, what about a "priority" decorator? Either `@priority(5)` or `[priority 5]` and so on

How would you do this in script modules meant to be used in multiple games, or as libraries in other people games? With these decorators would you have to adjust it every time you add one to a game?

**Comment by ericoporto** [comment]
Yeahp, just realized I have no idea how to externate this - so that the module user can adjust the order to their need.

**Comment by tarnos12** [comment]
Reviving this thread to give it some attention.

Afaik AGS4(and I assume 3.6) already has the ability to use functions regardless of the order in which they are created.

But the issue is still there for structs, so the other in which they are defined matters and you can't use a struct that is defined lower down the script.

**Comment by ivan-mogilko** [comment]
> Afaik AGS4(and I assume 3.6) already has the ability to use functions regardless of the order in which they are created.

No, 3.6 does not have this change, it belongs only to the new compiler in ags4.

**Comment by ericoporto** [comment]
Is it possible to do `import struct MyStruct;` and then use it and define it at a later time?

**Comment by fernewelten** [comment]
@ericoporto No “import” because `struct MyStruct` is a type declaration, not a variable declaration. Types aren't exported nor imported, only variables can be. But it is possible to forward declare a struct:

After you go `managed struct MyStruct;`, you can declare ´Mystruct *var;`. Then the variable `var` can point to a record, and the compiler doesn't know yet what fields the record has. So `var = new MyStruct;` will yield an error message. Later on, you'll probably want to declare `MyStruct` completely and specify the fields, and from that point onwards, you can do `new MyStruct`.

**Comment by fernewelten** [comment]
The "old" compiler has forward declarations, too.

@tarnos12 
> But the issue is still there for structs, so the order in which they are defined matters and you can't use a struct that is defined lower down the script.

Yes, unfortunately this is only implemented for functions, as a special case. We __could__ do this for types, too, in principle, but it would become very complicated. For instance, the struct type could have components that are declared even later in their turn. There might be declarations that are mutually dependent on each other, and so on. 

You __can__ declare pointers to struct without specifying what fields a struct has, the way `C` does it. See my posting above.

**Comment by tarnos12** [comment]
@fernewelten

I will keep that in mind, but it only works with managed structs.

Since it's very complicated then it's probably not worth the effort.

**Comment by fernewelten** [comment]
Yes, I would think that you can only sensibly forward-declare __managed__ structs. That is, you might be allowed to forward-declare a __non-managed__ struct – I don't know off top of my head whether the compiler will stop you. That is, you might go `struct Foo;`. But the trouble is, you can't __do__ anything with that sort of incomplete declaration:

Wherever you go `Foo var;`, the compiler will have to allocate enough bytes for `var`. And at this point in code, the compiler doesn't know yet how many bytes are enough to allocate, because the compiler doesn't know yet what components `Foo` has. 

On the other hand, when you go `managed struct Foo;` and later on `Foo *var`, then the compiler will only have to allocate enough bytes for a pointer at that point. That's something that the compiler can do even if the compiler doesn't know anything about `Foo` yet other than that `Foo` is a struct type.
